### PR TITLE
feat: update widget with client side SOR

### PR DESCRIPTION
**Pull request by ianlapham**
- replace sub-optimal client side routing with SOR on widget 
- create new `useBestTrade` hook that returns fallback route if SOR package fails

### CODE REVIEW

**Review by justindomingue — APPROVED** [review]

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        data: GetQuoteResult
+        error?: unknown
+      }
+    | undefined
+  >()
+
+  useEffect(() => {
+    async function fetchQuote() {
+      if (queryArgs) {
+        const result = await getClientSideQuote(queryArgs)
```
[the remark this answers is not in the document]
Comment (ianlapham) [reply src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:46]: Yeah good idea, will need error state to use fallback

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts [hunk 2]
```diff
@@ -43,12 +43,20 @@ export default function useClientSideSmartOrderRouterTrade<TTradeType extends Tr
   useEffect(() => {
     async function fetchQuote() {
       if (queryArgs) {
```
Comment (justindomingue) [inline src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:45]: Hm if query args isn’t defined you may never reset to loading(false)?might be better to set loading right before the await get client side quote ?

File: package.json [hunk 3]
```diff
@@ -162,7 +161,7 @@
     "@ethersproject/constants": "^5.4.0",
     "@ethersproject/contracts": "^5.4.1",
     "@ethersproject/hash": "^5.4.0",
-    "@ethersproject/providers": "^5.4.5",
+    "@ethersproject/providers": "5.4.0",
```
Comment (JFrankfurt) [inline package.json:164]: version drop intended?

File: src/lib/hooks/routing/clientSideSmartOrderRouter/index.ts [hunk 4]
```diff
@@ -50,3 +51,49 @@ export async function getQuote(
...
+  tokenInChainId: ChainId
+  tokenInDecimals: number
+  tokenInSymbol?: string
+  tokenOutAddress: string
+  tokenOutChainId: ChainId
+  tokenOutDecimals: number
+  tokenOutSymbol?: string
+  amount: string
+  type: 'exactIn' | 'exactOut'
+}) {
```
Comment (JFrankfurt) [inline src/lib/hooks/routing/clientSideSmartOrderRouter/index.ts:78]: can you move this up to a type definition?

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      if (queryArgs) {
+        try {
+          const result = await getClientSideQuote(queryArgs)
+          setLoading(false)
+          setFetchedResult({
+            data: result.data,
+            error: result.error,
+          })
+        } catch (e) {
+          setLoading(false)
```
Comment (JFrankfurt) [inline src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:54]: nit: use `finally` for `setLoading(false)`

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    const otherAmount =
+      tradeType === TradeType.EXACT_INPUT
+        ? currencyOut && quoteResult
+          ? CurrencyAmount.fromRawAmount(currencyOut, quoteResult.quote)
+          : undefined
+        : currencyIn && quoteResult
+        ? CurrencyAmount.fromRawAmount(currencyIn, quoteResult.quote)
+        : undefined
```
Comment (JFrankfurt) [inline src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:101]: My preference is to avoid nested ternaries in favor of `if (variableDescribingTheSituation) { .. }`

File: src/lib/hooks/routing/clientSideSmartOrderRouter/index.ts (hunk 4 above)
Comment (JFrankfurt) [reply src/lib/hooks/routing/clientSideSmartOrderRouter/index.ts:78]: there are a few function signatures here that I think also merit their own type definitions

**Review by JFrankfurt — APPROVED** [review]
A few nitpicky things

File: package.json (hunk 3 above)
Comment (zzmp) [reply package.json:164]: did you talk to SOR authors about this version drop? might be better for them to bump than for you to drop.
Alternatively, they should be using `@ethersproject` and not `ethers`, and using `^` (at-least) versioning, so that the projects can coalesce on one version.

File: src/lib/hooks/swap/useSwapInfo.tsx [hunk 7]
```diff
@@ -55,10 +55,7 @@ function useComputeSwapInfo(): SwapInfo {
     [inputCurrency, isExactIn, outputCurrency, amount]
   )
 
-  /**
-   * @TODO (ianlapham): eventually need a strategy for routing API here
```
Comment (zzmp) [inline src/lib/hooks/swap/useSwapInfo.tsx:59]: Is this comment resolved?

File: yarn.lock [hunk 8]
```diff
@@ -1828,6 +1828,31 @@
   dependencies:
     "@ethersproject/logger" "^5.5.0"
 
+"@ethersproject/providers@5.4.0":
```
Comment (zzmp) [inline yarn.lock:1831]: Run `npx yarn-deduplicate` to remove the extra copy. (This should be it's own commit.)

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        } catch (e) {
+          setLoading(false)
+          setFetchedResult({
+            data: undefined,
+            error: true,
+          })
+        }
+      }
+    }
+    setLoading(true)
```
Comment (zzmp) [inline src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:62]: nit: move these calls above the definition and separate with a newline. the definiton will be hoisted, and IMO it's more readable that way.

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    fetchQuote()
+  }, [queryArgs])
+
+  const [quoteResult, error] = useMemo(() => {
```
Comment (zzmp) [inline src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:66]: This is unnecessary - `fetchedResult` should take this typing initially.

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts (hunk 2 above)
Comment (zzmp) [reply src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:45]: or wrap the whole thing in a `try {} finally {}`

File: src/lib/hooks/routing/useRoutingAPIArguments.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
+import { useMemo } from 'react'
+
+/**
+ * Returns query arguments for the Routing API query or undefined if the
+ * query should be skipped.
```
Comment (zzmp) [inline src/lib/hooks/routing/useRoutingAPIArguments.ts:6]: Include in the doc comment that the input does not need to be memoized, as it will be destructured.

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+  const [fetchedResult, setFetchedResult] = useState<
+    | {
+        data: GetQuoteResult | undefined
+        error?: unknown
+      }
+    | undefined
+  >()
+
+  useEffect(() => {
+    async function fetchQuote() {
```
Comment (zzmp) [inline src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:44]: If the only thing this depends on is `queryArgs`, move the definition outside of the component, and pass in `queryArgs` explicitly, to avoid the inline definition and long component definition.

File: package.json (hunk 3 above)
Comment (ianlapham) [reply package.json:164]: yeah we need this version to match the version from SOR package for types to be compatiable

File: src/lib/hooks/swap/useSwapInfo.tsx (hunk 7 above)
Comment (ianlapham) [reply src/lib/hooks/swap/useSwapInfo.tsx:59]: Kinda depends, id say its more of a product decision now if we want to use the API, but now that we have SOR we aren't missing anything significant - ill leave a note that this is where routing API logic would go

File: src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts (hunk 12 above)
Comment (ianlapham) [reply src/lib/hooks/routing/useClientSideSmartOrderRouterTrade.ts:44]: Would need to accept the state setters too, is that weird?

**Review by JFrankfurt — APPROVED** [review]
